server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropStore.java

```diff
@@ -0,0 +1,397 @@
+    PropCacheId sysPropsId = PropCacheId.forSystem(context.getInstanceID());
+    return createInitialProps(context, sysPropsId, initProps);
+  }
+
+  /**
+   * Create initial properties if they do not exist. If the node exists, initialization will be
+   * skipped.
+   *
+   * @param context
+   *          the system context
+   * @param propCacheId
+   *          a prop id
+   * @param props
+   *          initial properties
+   * @return true if props create, false if the node existed and initialization was skipped.
+   */
+  public static boolean createInitialProps(final ServerContext context,
+      final PropCacheId propCacheId, Map<String,String> props) {
+
+    try {
+      ZooReaderWriter zrw = context.getZooReaderWriter();
+      if (zrw.exists(propCacheId.getPath())) {
+        return false;
+      }
+      VersionedProperties vProps = new VersionedProperties(props);
+      return zrw.putPersistentData(propCacheId.getPath(),
+          context.getVersionedPropertiesCodec().toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL);
+    } catch (InterruptedException ex) {
+      Thread.currentThread().interrupt();
+      throw new IllegalStateException("Interrupted creating node " + propCacheId, ex);
+    } catch (Exception ex) {
+      throw new IllegalStateException("Failed to create node " + propCacheId, ex);
+    }
+  }
+
+  public PropStoreMetrics getMetrics() {
+    return cacheMetrics;
+  }
+
+  // TODO - evaluate returning the props instead of boolean.
+  @Override
+  public boolean create(PropCacheId propCacheId, Map<String,String> props)
+      throws PropStoreException {
+
+    VersionedProperties vProps = new VersionedProperties(props);
+
+    try {
+
+      var path = propCacheId.getPath();
+      if (!zrw.putPersistentData(path, propCodec.toBytes(vProps), ZooUtil.NodeExistsPolicy.FAIL)) {
+        return false;
+      }
+
+      Stat stat = zrw.getStatus(path, propStoreWatcher);
+
+      if (stat.getVersion() != vProps.getNextVersion()) {
+        throw new PropStoreException(
+            "Invalid data version on create, received " + stat.getVersion(),
+            new IllegalStateException());
+      }
+
+      return true;
+
+    } catch (IOException | KeeperException | InterruptedException ex) {
+      throw new PropStoreException("Failed to serialize properties for " + propCacheId, ex);
+    }
+  }
+
+  @Override
+  public VersionedProperties get(final PropCacheId propCacheId) throws PropStoreException {
+    try {
+
+      checkZkConnection(); // if ZK not connected, block, do not just return a cached value.
+      propStoreWatcher.registerListener(propCacheId, this);
+      return cache.get(propCacheId);
+
+    } catch (Exception ex) {
+      throw new PropStoreException("read from prop store get() failed for: " + propCacheId, ex);
+    }
+  }
+
+  @Override
+  public boolean putAll(final PropCacheId propCacheId, final Map<String,String> props)
+      throws PropStoreException {
+
+    try {
+
```

Comment:
I had thought about allowing retries on a conflict, but I could not reason out how to handle a conflict.  If both requests changed different things, then maybe it would be okay.

```
Thread-1 sets prop A to x
Thread-2 sets prop B to y

this seems "safe"
```

But what if it was a change?
```
prop A = z.
Thread-1 sets prop A = x
Thread-2 sets prop A = y
who wins?  If both return success, then what is the the value of prop A after both return?  Would you need to set and then call read to see if you "won"?
````

Rather than try to figure out logic that would be the least surprising, it seemed easiest to hard / quick fail and let the caller decide.  The calling function could make the decision to retry - it should have more context available to decide on what should happen in the case of a conflict.

Reply:
For the case of two thread setting the same thing concurrently we want to achieve the same result as them running serially in any order.  For the concurrent case if one succeeds and the other fails and retries, then we should get the same result as them running serially.

For example the following

```
Thread-1 reads the current props and stat from ZK
Thread-2 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-2 modified the props in memory to set A=y,C=q
Thread-1 successfully writes the modified props to ZK
Thread-2 fails to write the props to ZK because the version does not match what it read
Thread-2 reads the current props and stat from ZK
Thread-2 modifies the props to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
```

The above sequence of events should result in the props in ZK being `A=y,B=z,C=q` which is the same as what we would see if there was no concurrency in the following set of events. Neither one of these outcomes are surprising. 

```
Thread-1 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-1 successfully writes the modified props to ZK
Thread-2 reads the current props and stat from ZK
Thread-2 modifies the props in memory to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
```

Reply:
But what about:
```
Thread-1 reads the current props and stat from ZK
Thread-2 reads the current props and stat from ZK
Thread-1 modifies the props in memory to set A=x,B=z
Thread-2 modified the props in memory to set A=y,C=q
Thread-2 successfully writes the modified props to ZK
Thread-1 fails to write the props to ZK because the version does not match what it read
Thread-1 reads the current props and stat from ZK
Thread-1 modifies the props to set A=x,C=z
Thread-2 successfully writes the modified props to ZK
```
This would result in `A=x, B=z, C=q` not `A=y, B=z, C=q` in this case I would think that we would not want to retry without somehow evaluating if A=x or A=y is the desired value.

Reply:
> This would result in A=x, B=z, C=q not A=y, B=z, C=q in this case I would think that we would not want to retry without somehow evaluating if A=x or A=y is the desired value.

That is fine ` A=x, B=z, C=q` and `A=y, B=z, C=q` are both acceptable outcomes.  We want to achieve the same result as the operations running serially in any order.  Unacceptable outcomes of concurrency would be final zookeeper states of `A=x,B=z` or `A=y,C=q` because there is no serial ordering that would achieve those.

Reply:
For a more complex example if the following three set operations run concurrently.
  
 * `A=x,B=z`
 * `B=t,C=s`
 * `A=y,C=q`
 
Then the following is all of the ways these operations could run serially, the result of running them concurrently must result in one of these outcomes to be considered correct.

 * `A=x,B=z` then `B=t,C=s` then   `A=y,C=q` yielding `A=y,B=t,C=q`
 * `A=x,B=z` then `A=y,C=q` then `B=t,C=s`  yielding `A=y,B=t,C=s`
 * `A=y,C=q` then `A=x,B=z` then `B=t,C=s`  yielding `A=x,B=t,C=s`
 * `A=y,C=q` then `B=t,C=s` then `A=x,B=z`   yielding `A=x,B=z,C=s`
 * `B=t,C=s` then `A=y,C=q` then  `A=x,B=z`   yielding `A=x,B=z,C=q`
 * `B=t,C=s` then `A=x,B=z` then `A=y,C=q`    yielding `A=y,B=z,C=q`

Reply:
I disagree that any of them would be "correct". From an operation ordering viewpoint they are all consistent - but I argue that one of them represents the desired state.

As a contrived example, say two people use the shell and try to set the split threshold on a table, one picks 512M and the other uses 1G.  One of those values should be more correct than the other - which one would be left for them to figure out.  If both submitted the change such that they run concurrently, whatever the order that the operations run, one value will be the ultimate value of the property.  What I think you are suggesting is that both would receive success and neither would know which value was set without double checking.

For this level of the code - I'm proposing that one should succeed and the other should fail.  The one that receives the failure then has the option to either accept the change that got in first and "won", or they resubmit and change it to what they want.

The decision would be taken outside of this code - some operations may not care and elect to retry - others may choose throw an error - but it also leaves open other possibilities - maybe non-conflicting changes can be merged. That is, one sets X and the other sets Y, then setting both X and Y is acceptable. I didn't what to try and have that arbitration occur at this level.

Reply:
> What I think you are suggesting is that both would receive success and neither would know which value was set without double checking.

That situation can happen even w/o operations to set a property overlapping in time.  Consider the following.

 * via the shell user A makes an Accumulo API call to set the split threshold to 1G.  This call completes successfully w/o any other API set prop calls running concurrently.  The API call returns true and the shell stores this return value in a variable for user A's shell instance. 
 * via the shell user B makes an Accumulo API call to set the split threshold to 512M.  This call completes successfully w/o any other API set prop calls running concurrently.  The API call returns true and the shell stores this return value in a variable for user B's shell instance.
 * User A' shell prints success.
 * User B's shell prints success.

So even if the API code failed when concurrency was detected, it would not fail in this case because the concurrency happened outside the API call.  However the concurrency still happened and one user's set was overwritten by the other even though success was reported.

Reply:
The only way to deal the concurrency problem at the Accumulo API level would be to expose something like the version for getting and setting properties. The current APIs should probably retry.  If we have new APIs that expose this information, then those could fail if the version changed since reading the properties.

Reply:
Another way to deal with concurrency at the API level instead of exposing version information would be a new API that supports conditional set.  An API where a user could do something like set  `A=x,B=z` only if `A=r,B=y,C=m`.  This could be implemented on the backend using the version information w/o exposing the version information in the API.

Reply:
This code is not public API and right now is strongly geared to configuration. I'd like to propose that we defer on this until the next layer PR is reviewed.  The next layer implements the configuration and if in that context enough calls would be simplified and benefit from just retrying I'll implement the retry here with the mutator.

Reply:
>  The next layer implements the configuration and if in that context enough calls would be simplified and benefit from just retrying I'll implement the retry here with the mutator.

That's fine.  The discussion was good, I learned a few things from it.  Mainly I realized that our current API only supports unconditional set operations that do not care about any previous state.  Unconditional set operations should just retry when there is a failure because of concurrency because they have no concept of a required precondition. If we want to support conditional set operations, then those pre-conditions would have to be represented in the API in some way.